com.unity.multiplayer.mlapi/Runtime/Core/NetworkUpdateLoop.cs

```diff
@@ -0,0 +1,525 @@
+                        {
+                            Debug.LogError($"{nameof(NetworkUpdateLoop)}.{nameof(Initialize)}: Cannot find index of `{nameof(Update.ScriptRunBehaviourUpdate)}` loop system in `{nameof(Update)}`'s subsystem list!");
+                            return;
+                        }
+
+                        // insert before `Update.ScriptRunBehaviourUpdate`
+                        subsystems.Insert(indexOfScriptRunBehaviourUpdate, NetworkUpdate.CreateLoopSystem());
+                    }
+                    playerLoopSystem.subSystemList = subsystems.ToArray();
+                }
+                else if (playerLoopSystem.type == typeof(PreLateUpdate))
+                {
+                    var subsystems = playerLoopSystem.subSystemList.ToList();
+                    {
+                        int indexOfScriptRunBehaviourLateUpdate = subsystems.FindIndex(s => s.type == typeof(PreLateUpdate.ScriptRunBehaviourLateUpdate));
+                        if (indexOfScriptRunBehaviourLateUpdate < 0)
+                        {
+                            Debug.LogError($"{nameof(NetworkUpdateLoop)}.{nameof(Initialize)}: Cannot find index of `{nameof(PreLateUpdate.ScriptRunBehaviourLateUpdate)}` loop system in `{nameof(PreLateUpdate)}`'s subsystem list!");
+                            return;
+                        }
+
+                        // insert before `PreLateUpdate.ScriptRunBehaviourLateUpdate`
+                        subsystems.Insert(indexOfScriptRunBehaviourLateUpdate, NetworkPreLateUpdate.CreateLoopSystem());
+                    }
+                    playerLoopSystem.subSystemList = subsystems.ToArray();
+                }
+                else if (playerLoopSystem.type == typeof(PostLateUpdate))
+                {
+                    var subsystems = playerLoopSystem.subSystemList.ToList();
+                    {
+                        int indexOfPlayerSendFrameComplete = subsystems.FindIndex(s => s.type == typeof(PostLateUpdate.PlayerSendFrameComplete));
+                        if (indexOfPlayerSendFrameComplete < 0)
+                        {
+                            Debug.LogError($"{nameof(NetworkUpdateLoop)}.{nameof(Initialize)}: Cannot find index of `{nameof(PostLateUpdate.PlayerSendFrameComplete)}` loop system in `{nameof(PostLateUpdate)}`'s subsystem list!");
+                            return;
+                        }
+
+                        // insert after `PostLateUpdate.PlayerSendFrameComplete`
+                        subsystems.Insert(indexOfPlayerSendFrameComplete + 1, NetworkPostLateUpdate.CreateLoopSystem());
+                    }
+                    playerLoopSystem.subSystemList = subsystems.ToArray();
+                }
+
+                customPlayerLoop.subSystemList[i] = playerLoopSystem;
+            }
+
+            PlayerLoop.SetPlayerLoop(customPlayerLoop);
+        }
+
+        private struct NetworkInitialization
```

Comment:
I find the use of multiple classes (`NetworkInitialization`, `NetworkEarlyUpdate`, `NetworkFixedUpdate`, ...) to be very class-heavy. Since the class don't have different functions and members, it seems a single class could do with a parameter, maybe of type `NetworkUpdateStage`.

Reply:
Seems like this would be less error prone / more extensible / easier to read...but @MFatihMAR I thought you had a rationale for going this way?

Reply:
they are not exposed and they don't implement anything other than `CreateLoopSystem`.

2 main reasons:

1. it's Unity-like: [`EarlyUpdate.ScriptRunDelayedStartupFrame`](https://docs.unity3d.com/2019.4/Documentation/ScriptReference/PlayerLoop.EarlyUpdate.ScriptRunDelayedStartupFrame.html)
2. it allows us to differentiate between `PlayerLoopSystem`s by `type = typeof(XXX)` (that's also how we find specific places to inject our stuff)

so, I'd say they serve a pretty valid purpose here.

Reply:
Ok, but why not (similar to my other comment) declare a dictionary with key of `PostLateUpdate`, etc. and they the value being the function?